File: packages/@xrnjs/multi-bundle/harmony/xrn_multi_bundle/src/main/ets/bundle/BundleInfo.ts
import { BundleInfoHook } from "./BundleInfoManager"

/**
 * Bundle 信息
 * 静态数据
 */
export class BundleInfo {

  static BUNDLE_TYPE_MAIN = "main"
  static BUNDLE_TYPE_REMOTE = "remote"

  static DELIVERY_TYPE_INNER = "INNER"
  static DELIVERY_TYPE_DYNAMIC = "DYNAMIC"
  static DELIVERY_TYPE_LOCAL = "LOCAL"

  static DYNAMIC_UNKNOWN_PORT = 1

  private static TAG = "BundleInfo"

  /**
   * bundle name
   */
  readonly bundleName: string = ""
  /**
   * 是否主bundle
   */
  readonly bundleType: string = ""
  /**
   * module 名，main bundle 需要配置
   */
  readonly defaultModuleName: string = ""
  /**
   * 所有的 moduleName
   */
  private readonly moduleNames: string[] = []
  /**
   * CodePush Key
   */
  private readonly codePushKey: string = ""
  /**
   * 本地服务端口
   */
  private port: number = 8081

  private readonly deliveryType: string = BundleInfo.DELIVERY_TYPE_INNER

  /**
   * hook
   */
  private hook: BundleInfoHook | undefined

  constructor(bundleName: string, bundleType: string, defaultModuleName: string, moduleNames: string[], codePushKey: string, port: number, deliveryType?: string) {
    if (bundleName.length == 0) {
      throw new Error(`${BundleInfo.TAG}.constructor:bundleName is empty`)
    }
    this.bundleName = bundleName
    this.bundleType = bundleType
    this.defaultModuleName = defaultModuleName;
    this.moduleNames = moduleNames || [];
    this.codePushKey = codePushKey
    this.port = port
    this.deliveryType = deliveryType
  }

  addAppKey(appKey: string | undefined) {
    if (!appKey) {
      return;
    }
    if (!this.moduleNames.includes(appKey)) {
      this.moduleNames.push(appKey);
    }
  }

  getModuleNames(): string[] {
    return [...this.moduleNames];
  }

  setHook(hook: BundleInfoHook | undefined) {
    this.hook = hook
  }

  /**
   * 是否为 main bundle
   */
  isMainBundle(): boolean {
    return this.bundleType === BundleInfo.BUNDLE_TYPE_MAIN
  }

  /**
   * 返回初始配置的CodePushKey
   * @returns
   */
  getInitCodePushKey(): string {
    return this.codePushKey;
  }

  /**
   * 获取 CodePush key
   * @returns
   */
  getCodePushKey(): string {
    if (this.hook?.hookCodePushKey) {
      return this.hook.hookCodePushKey?.(this) || ""
    } else {
      return this.codePushKey
    }
  }

  /**
   * 获取本地服务端口
   * @returns
   */
  getLocalServerPort(): number {
    if (this.hook?.hookLocalServerPort) {
      return this.hook.hookLocalServerPort(this, this.port)
    }
    return this.port
  }

  /**
   * 设置端口
   * 主要用于 LocalBundle 设置
   * @param port
   */
  setLocalServerPort(port: number) {
    this.port = port;
  }

  /**
   * 获取 js bundle 文件名
   * @returns
   */
  getJSBundleName(): string {
    return this.hook?.hookJSBundleName?.(this) || `oh.${this.bundleName}.bundle`
  }

  /**
   * 获取 本地服务 url
   * @returns
   */
  getLocalServerUrl(): string {
    return this.hook?.hookLocalServerUrl?.(this) || `http://localhost:${this.getLocalServerPort()}/index.bundle?platform=harmony&dev=true&minify=false`
  }

  getDeliveryType(): string {
    return this.deliveryType
  }
}

File: packages/@xrnjs/multi-bundle/android/src/main/java/xrn/modules/multibundle/bundle/BundleInfo.kt
package xrn.modules.multibundle.bundle

import com.facebook.react.BuildConfig
import xrn.modules.multibundle.devsupport.XRNDeveloperSettings


/**
 * Bundle 信息
 * 静态数据
 * @param bundleName bundle 名
 * @param bundleType bundle 类型
 * @param defaultModuleName 默认 module 名
 * @param moduleNames module 名数组
 * @param codePushKey CodePush Key
 * @param port 端口
 */
class BundleInfo(
    val bundleName: String,
    val bundleType: BundleType,
    val defaultModuleName: String,
    moduleNames: Array<String>?,
    codePushKey: String?,
    private var port: Int,
    private val deliveryType: DeliveryType = DeliveryType.INNER
) {

    companion object {
        /**
         * main bundle
         */
        const val BUNDLE_TYPE_MAIN = "main"
    }

    /**
     * 初始 CodePush Key
     */
    private val initCodePushKey: String = codePushKey ?: ""

    /**
     * module name list
     */
    private val moduleNameList = mutableListOf<String>()

    init {
        if (!moduleNames.isNullOrEmpty()) {
            moduleNameList.addAll(moduleNames)
        }
    }

    /**
     * Hook
     */
    private var hook: BundleInfoHook? = null

    /**
     * 是否为主 bundle
     */
    fun isMainBundle(): Boolean {
        return bundleType == BundleType.MAIN
    }

    /**
     * 添加 AppKey
     */
    fun addAppKey(appKey: String?) {
        if (appKey.isNullOrEmpty()) {
            return
        }
        if (!moduleNameList.contains(appKey)) {
            moduleNameList.add(appKey)
        }
    }

    /**
     * 获取 module 名 list
     */
    fun getModuleNames(): List<String> {
        return moduleNameList.toList()
    }

    /**
     * 设置 BundleInfoHook
     */
    fun setHook(hook: BundleInfoHook?) {
        this.hook = hook
    }

    /**
     * 获取初始 CodePush Key
     */
    fun getInitCodePushKey(): String {
        return initCodePushKey
    }

    /**
     * 获取 CodePush Key
     */
    fun getCodePushKey(): String {
        if (BuildConfig.DEBUG
            && XRNDeveloperSettings.instance(bundleName)?.isCodePushEnabled() == false
        ) {
            return ""
        }

        return hook?.hookCodePushKey?.let { it(this) } ?: initCodePushKey
    }

    /**
     * 获取本地服务 端口
     */
    fun getPort(): Int {
        val cachedPort = XRNDeveloperSettings.instance(bundleName)?.getDebugServerPort()
        if (cachedPort != null && cachedPort != XRNDeveloperSettings.DEBUG_SERVER_PORT_DEFAULT) {
            return cachedPort
        }

        return port
    }

    fun setPort(newPort: Int) {
        this.port = newPort
    }
}

enum class BundleType {
    MAIN,
    DEFAULT
}

enum class DeliveryType {
    INNER,
    DYNAMIC
}

/**
 * 特殊信息Hook方法
 */
interface BundleInfoHook {
    val hookCodePushKey: ((info: BundleInfo) -> String)?
}
